text/0059-nodes-capabilities-discovery.md

```diff
@@ -0,0 +1,142 @@
+
+This RFC attempts to solve this problem by giving the possibility to build a list of nodes that are capable of serving specific data.
+
+## Stakeholders
+
+Low-level client developers.
+People interested in accessing the archive of the chain.
+
+## Explanation
+
+*Reading RFC #8 first might help with comprehension, as this RFC is very similar.*
+
+Please keep in mind while reading that everything below applies for both relay chains and parachains, except mentioned otherwise.
+
+### Capabilities
+
+This RFC defines a list of so-called **capabilities**:
+
+- **Head of chain provider**. An implementation with this capability must be able to serve to other nodes block headers, block bodies, justifications, calls proofs, and storage proofs of "recent" (see below) blocks, and, for relay chains, to serve to other nodes warp sync proofs where the starting block is a session change block and must participate in Grandpa and Beefy gossip.
+- **History provider**. An implementation with this capability must be able to serve to other nodes block headers and block bodies of any block since the genesis, and must be able to serve to other nodes justifications of any session change block since the genesis up until and including their currently finalized block.
+- **Archive provider**. This capability is a superset of **History provider**. In addition to the requirements of **History provider**, an implementation with this capability must be able to serve call proofs and storage proof requests of any block since the genesis up until and including their currently finalized block.
+- **Parachain bootnode** (only for relay chains). An implementation with this capability must be able to serve the network request described in RFC 8.
+
+In the context of the *head of chain provider*, the word "recent" means: any not-finalized-yet block that is equal to or an ancestor of a block that it has announced through a block announce, and any finalized block whose height is superior to its current finalized block minus **16**.
+This does *not* include blocks that have been pruned because they're not a descendant of its current finalized block. In other words, blocks that aren't a descendant of the current finalized block can be thrown away.
+A gap of blocks is required due to race conditions: when a node finalizes a block, it takes some time for its peers to be made aware of this, during which they might send requests concerning older blocks. The exact gap is arbitrary.
+
+Substrate is currently by default a **head of chain provider** provider. After it has finished warp syncing, it downloads the list of old blocks, after which it becomes a **history provider**.
+If Substrate is instead configured as an archive node, then it downloads the state of all blocks since the genesis, after which it becomes an **archive provider**, **history provider**, and **head of chain provider**.
```

Comment:
point of information: though it doesn't matter for the purposes of this PR how the archive data is acquired, a typical node computes the incremental states rather than downloading them, no?

Reply:
Yes, that's maybe a poor choice of wording. I used the word "download" because it downloads the block headers and bodies.